ticdc: add output-raw-change-event parameter and update ticdc-behavior-change (#18170)
pingcap/docs: ticdc/ticdc-split-update-behavior.md

> **Note:**
>
> After this behavior change, when using the MySQL sink, TiCDC does not split the `UPDATE` event in most cases. Consequently, there might be primary key or unique key conflicts during changefeed runtime, causing the changefeed to restart automatically. After the restart, TiCDC will split the conflicting `UPDATE` events into `DELETE` and `INSERT` events before writing them to the Sorter module. This ensures that all events within the same transaction are correctly ordered, with all `DELETE` events preceding `INSERT` events, thus correctly completing data replication.

> ## Split `UPDATE` primary or unique key events for non-MySQL sinks

Author: ```suggestion
## Split primary or unique key `UPDATE` events for non-MySQL sinks
```

### Transactions containing multiple `UPDATE` changes

Starting from v6.5.4, v7.1.2, and v7.4.0, for transactions containing multiple changes, if the primary key or non-null unique index value is modified in the `UPDATE` event, TiCDC splits the event into `DELETE` and `INSERT` events and ensures that all events follow the sequence of `DELETE` events preceding `INSERT` events. For more information, see GitHub issue [#9430](https://github.com/pingcap/tiflow/issues/9430).

This change primarily addresses the potential issue of primary key or unique key conflicts when using the MySQL sink to directly write these two events to the downstream, leading to changefeed errors. When using the Kafka sink or other sinks, you might encounter the same error if the consumer writes messages to a relational database or performs similar operation.

Take the following SQL as an example:

```sql
CREATE TABLE t (a INT PRIMARY KEY, b INT);
INSERT INTO t VALUES (1, 1);
INSERT INTO t VALUES (2, 2);

BEGIN;
UPDATE t SET a = 3 WHERE a = 1;
UPDATE t SET a = 1 WHERE a = 2;
UPDATE t SET a = 2 WHERE a = 3;
COMMIT;
```

> In this example, by executing three SQL statements to swap the primary keys of two rows, TiCDC only receives two update change events, that is, changing the primary key `a` from `1` to `2` and changing the primary key `a` from `2` to `1`. If the MySQL sink directly writes these two `UPDATE` events to the downstream, a primary key conflict might occur, leading to changefeed errors.

Reviewer: ```suggestion
In this example, by executing three SQL statements to swap the primary keys of two rows, TiCDC only receives two update change events, that is, changing the primary key `a` from `1` to `2` and changing the primary key `a` from `2` to `1`. If the MySQL sink directly writes these two `UPDATE` events to the downstream, a primary key conflict will inevitably occur, leading to changefeed errors.
```

Therefore, TiCDC splits these two events into four events, that is, deleting records `(1, 1)` and `(2, 2)` and writing records `(2, 1)` and `(1, 2)`.

> ### Control whether to split `UPDATE` primary or unique key events

Author: ```suggestion
### Control whether to split primary or unique key `UPDATE` events
```

> Starting from v6.5.10, v7.1.6, v7.5.2, and v8.1.1, when using a non-MySQL sink, TiCDC supports controlling whether to split `UPDATE` primary or unique key events via the `output-raw-change-event` parameter, as described in the GitHub issue [#11211]( https://github.com/pingcap/tiflow/issues/11211). The specific behavior of this parameter is as follows:

Author: ```suggestion
Starting from v6.5.10, v7.1.6, v7.5.3, and v8.1.1, when using a non-MySQL sink, TiCDC supports controlling whether to split `UPDATE` primary or unique key events via the `output-raw-change-event` parameter, as described in the GitHub issue [#11211]( https://github.com/pingcap/tiflow/issues/11211). The specific behavior of this parameter is as follows:
```

- When you set `output-raw-change-event = false`, if the primary key of an `UPDATE` event or the value of a column of a non-null unique index is changed, TiCDC splits the event into two events, `DELETE` and `INSERT`, and ensures that all events are sorted in the order in which the `DELETE` event precedes the `INSERT` event.
- When you set `output-raw-change-event = true`, TiCDC does not split the `UPDATE` event. Note that when the primary key of a table is a clustered index, updates to the primary key are split into `DELETE` and `INSERT` events in TiDB, and such behavior is not affected by the `output-raw-change-event` parameter.

#### Release 6.5 compatibility

> | Version | Protocol | Split ALL UPDATE UK/PK | Not Split UPDATE UK/PK | Comments |

Author: ```suggestion
| Version | Protocol | Split All UK/PK `UPDATE` | Not Split UK/PK `UPDATE`  | Comments |
```

| -- | -- | -- | -- | -- |
| <= v6.5.2 | ALL | ✗ | ✓ |  |
| v6.5.3 / v6.5.4 | Canal/Open | ✗ | ✓ |  |
| v6.5.3 | CSV/Avro | ✗ | ✗ | Split but does not sort. See [#9086](https://github.com/pingcap/tiflow/issues/9658) |
| v6.5.4 | Canal/Open | ✗ | ✗ | Only split and sort transactions that contain multiple changes |
| v6.5.5 ～ v6.5.9 | ALL | ✓ | ✗ |
| \>= v6.5.10 | ALL | ✓ (Default value: `output-raw-change-event = false`) | ✓ (Optional: `output-raw-change-event = true`) | |

#### Release 7.1 compatibility

> | Version | Protocol | Split ALL UPDATE UK/PK | Not Split UPDATE UK/PK | Comments |

Author: ```suggestion
| Version | Protocol | Split All UK/PK `UPDATE` | Not Split UK/PK `UPDATE`  | Comments |
```

| -- | -- | -- | -- | -- |
| v7.1.0 | ALL | ✗ | ✓ |  |
| v7.1.1 | Canal/Open | ✗ | ✓ |  |
| v7.1.1 | CSV/Avro | ✗ | ✗ | Split but does not sort. See [#9086](https://github.com/pingcap/tiflow/issues/9658) |
| v7.1.2  ~ v7.1.5 | ALL | ✓ | ✗ |  |
| \>= v7.1.6 (not released yet) | ALL | ✓ (Default value: `output-raw-change-event = false`) | ✓ (Optional: `output-raw-change-event = true`)  | |

#### Release 7.5 compatibility

> | Version | Protocol | Split ALL UPDATE UK/PK | Not Split UPDATE UK/PK | Comments |

Author: ```suggestion
| Version | Protocol | Split All UK/PK `UPDATE` | Not Split UK/PK `UPDATE`   | Comments |
```

| -- | -- | -- | -- |
| <= v7.5.2 | ALL | ✓ | ✗ |
| \>= v7.5.3 (not released yet) | ALL | ✓ (Default value:`output-raw-change-event = false`) | ✓  (Optional: `output-raw-change-event = true`) | |

#### Release 8.1 compatibility

> | Version | Protocol | Split ALL UPDATE UK/PK | Not Split UPDATE UK/PK | Comments |

Author: ```suggestion
| Version | Protocol | Split All UK/PK `UPDATE` | Not Split UK/PK `UPDATE`  | Comments |
```
